src/common/cockpitwebinject.c

```diff
@@ -84,62 +87,68 @@ cockpit_web_inject_push (CockpitWebFilter *filter,
   const gchar *mark, *data, *pos;
   gsize mark_len, data_len, at;
   GBytes *bytes;
+  gsize written;
 
   /* Stop searching once injected */
-  if (self->inject)
+  mark = g_bytes_get_data (self->marker, &mark_len);
+  data = g_bytes_get_data (block, &data_len);
+  written = at = 0;
+
+  while (self->injected < self->maximum && at < data_len)
     {
-      mark = g_bytes_get_data (self->marker, &mark_len);
-      data = g_bytes_get_data (block, &data_len);
-      at = 0;
+      if (self->partial)
+        pos = data + at;
+      else
+        pos = memchr (data + at, mark[0], data_len - at);
 
-      while (at < data_len)
-        {
-          if (self->partial)
-            pos = data + at;
-          else
-            pos = memchr (data + at, mark[0], data_len - at);
+      /* Couldn't find the character anywhere? */
+      if (!pos)
+        break;
 
-          /* Couldn't find the character anywhere? */
-          if (!pos)
+      for (at = (pos - data); self->partial < mark_len && at < data_len; self->partial++, at++)
+        {
+          if (mark[self->partial] != data[at])
             break;
+        }
 
-          for (at = (pos - data); self->partial < mark_len && at < data_len; self->partial++, at++)
-            {
-              if (mark[self->partial] != data[at])
-                break;
-            }
+      /* Found a match */
+      if (self->partial == mark_len)
+        {
+          self->partial = 0;
 
-          /* Found a match */
-          if (self->partial == mark_len)
+          if (written < at)
             {
-              self->partial = 0;
-
-              bytes = g_bytes_new_from_bytes (block, 0, at);
+              bytes = g_bytes_new_from_bytes (block, written, at - written);
               function (func_data, bytes);
               g_bytes_unref (bytes);
+            }
 
-              function (func_data, self->inject);
-              g_bytes_unref (self->inject);
-              self->inject = NULL;
+          function (func_data, self->inject);
+          self->injected++;
 
-              bytes = g_bytes_new_from_bytes (block, at, data_len - at);
-              function (func_data, bytes);
-              g_bytes_unref (bytes);
-
-              block = NULL;
-              break;
-            }
+          written = at;
+        }
 
-          /* Incomplete match, and more data */
-          else if (at < data_len)
-            {
-              self->partial = 0;
-            }
+      /* Incomplete match, and more data */
+      else if (at < data_len)
+        {
+          self->partial = 0;
         }
     }
 
-  if (block)
-    function (func_data, block);
+  if (written < data_len)
+    {
+      if (written == 0)
+        {
+          function (func_data, block);
+        }
+      else
+        {
+          bytes = g_bytes_new_from_bytes (block, written, data_len - written);
```

Comment:
Throughout the function, data from `block` is allocated for and copied in chunks. Can the consumer receive a const pointer and we just get the data once? We keep the index anyway.
Then this `written == 0` case wouldn't need to be special either.